library/ssl_msg.c

```diff
@@ -1134,32 +1144,13 @@ static void mbedtls_ssl_cf_memcpy_if_eq( unsigned char *dst,
                                          size_t len,
                                          size_t c1, size_t c2 )
 {
-    /* diff = 0 if c1 == c2, non-zero otherwise */
-    const size_t diff = c1 ^ c2;
-
-    /* MSVC has a warning about unary minus on unsigned integer types,
-     * but this is well-defined and precisely what we want to do here. */
-#if defined(_MSC_VER)
-#pragma warning( push )
-#pragma warning( disable : 4146 )
-#endif
-
-    /* diff_msb's most significant bit is equal to c1 != c2 */
-    const size_t diff_msb = ( diff | -diff );
-
-    /* diff1 = (c1 != c2) in {0, 1} */
-    const size_t diff1 = diff_msb >> ( sizeof( diff_msb ) * 8 - 1 );
-
-    /* mask = c1 != c2 ? 0xff : 0x00 */
-    const unsigned char mask = (unsigned char) -diff1;
-
-#if defined(_MSC_VER)
-#pragma warning( pop )
-#endif
+    /* mask = c1 == c2 ? 0xff : 0x00 */
+    const size_t equal = mbedtls_ssl_cf_bool_eq( c1, c2 );
+    const unsigned char mask = mbedtls_ssl_cf_mask_from_bit( equal );
 
     /* dst[i] = c1 != c2 ? dst[i] : src[i] */
```

Comment:
Rewrite this due to the mask "inversion": "dst[i] = ( c1 == c2 ) ? src[i] : dst[i]" ? and change the code as well:
`dst[i] = ( src[i] & mask ) | ( dst[i] & ~mask )`.